File: 322.coin-change.cs
/*
 * @lc app=leetcode id=322 lang=csharp
 *
 * [322] Coin Change
 */

// @lc code=start
public class Solution
{
    public int CoinChange(int[] coins, int amount)
    {
        if (amount < 1) return 0;
        int[] result = new int[amount + 1];
        Array.Sort(coins);
        for (int i = 1; i <= amount; i++)
        {
            result[i] = int.MaxValue;
            foreach (var coin in coins)
            {
                if (coin > i) break;
                if (result[i - coin] != int.MaxValue)
                    result[i] = Math.Min(result[i], 1 + result[i - coin]);
            }
        }
        return result[amount] == int.MaxValue ? -1 : result[amount];
    }
}
// @lc code=end

File: 322.coin-change.js
/*
 * @lc app=leetcode id=322 lang=javascript
 *
 * [322] Coin Change
 */

// @lc code=start
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function (coins, amount) {
  if (amount === 0) return 0;
  let solution = Array(amount + 1).fill(Number.MAX_SAFE_INTEGER);
  solution[0] = 0;
  coins.sort(function (a, b) {
    return a - b;
  });
  for (let i = 1; i <= amount; i++) {
    for (let j = 0; j < coins.length; j++) {
      if (coins[j] > i) {
        j = coins.length;
        break;
      }
      if (solution[i - coins[j]] != Number.MAX_SAFE_INTEGER)
        solution[i] = Math.min(solution[i], solution[i - coins[j]] + 1);
    }
  }
  return solution[amount] == Number.MAX_SAFE_INTEGER ? -1 : solution[amount];
};
// @lc code=end
